File: Python/LinkedList.py


class Node(object):

    def __init__(self, data=None):
        self.data = data
        self.next = None

    # Represent data at this node
    def __repr__(self):
        return (
            repr(self.data)  # return this
            if self.data != None  # if
            else ""  # otherwise
        )


class LinkedList(object):

    def __init__(self):
        self.head = Node()

    # Length of linked list
    def __len__(self):
        size = 0
        for index, node in enumerate(self):
            # return 0 if no data within
            if node.data is None:
                return 0

            size = index + 1
        return size

    # Make linked list iterable

    def __iter__(self):
        node = self.head
        while node is not None:
            yield node
            node = node.next

    def insert(self, data):
        cur_node = self.head

        # traverse the list
        for node in self:
            # add to the first node
            # if the list is empty
            if node.data is None:
                self.head = Node(data)
                return
            elif node.next is None:
                cur_node = node
        # add to the last node
        # which is none
        cur_node.next = Node(data)

    def regain(self, data):
        result_set = []
        for index, node in enumerate(self):
            if node.data == data:
                found = {"index": index, "data": node}
                result_set.append(found)
        if len(result_set) > 0:
            return result_set
        else:
            raise Exception('Data cannot be found!')

    def alter(self, index, data):
        last_node = None
        for idx, node in enumerate(self):
            new_node = Node(data)
            if idx == index:
                if index == 0:
                    self.head = new_node
                    new_node.next = node.next
                    return
                else:
                    last_node.next = new_node
                    new_node.next = node.next
                    return
            last_node = node

    def delete(self, index):
        last_node = None
        for idx, node in enumerate(self):
            if idx == index:
                if index == 0:
                    self.head = node.next
                    return
                else:
                    last_node.next = node.next
                    return
            last_node = node


linked_list = LinkedList()
print(f"Init list: {list(linked_list)}\n")

linked_list.insert("Ed")
linked_list.insert("Ed")
linked_list.insert("Roy")
linked_list.insert("John")

print(f"After insertion: {list(linked_list)}\n")

data_set = linked_list.regain("Ed")
print(f"Regain data by given Ed var: {data_set}\n")

linked_list.alter(data_set[0]["index"], "Edward")
print(f"After alteration: {list(linked_list)}\n")

linked_list.delete(data_set[0]["index"])
print(f"After deletion of Ed: {list(linked_list)}\n")

File: Golang/pkg/data_structures/linked_list.go
package data_structures

type Node struct {
	value interface{}
	next  *Node
}

type LinkedList struct {
	head *Node
}

func (std *LinkedList) Insert(val interface{}) {
	if std.head == nil {
		std.head = &Node{val, nil}

		return
	}

	if std.head.next == nil {
		std.head.next = &Node{val, nil}

		return
	}

	next := std.head.next

	for next.next != nil {
		next = next.next
	}

	next.next = &Node{val, nil}
}

func (std *LinkedList) Retrieve(val interface{}) interface{} {
	var resultSet map[string]interface{}

	if std.head != nil {
		index := 0

		if std.head.value == val {
			resultSet = map[string]interface{}{
				"idx": index,
				"val": std.head.value,
			}
		}

		if resultSet == nil {
			next := std.head.next

			for next != nil {
				index++

				if next.value == val {
					resultSet = map[string]interface{}{
						"idx": index,
						"val": next.value,
					}
				}

				next = next.next
			}
		}
	}

	return resultSet
}

func (std *LinkedList) FilterByVal(val interface{}) interface{} {
	var resultSet []interface{}

	if std.head != nil {
		index := 0

		if std.head.value == val {
			resultSet = append(resultSet, map[string]interface{}{
				"idx": index,
				"val": std.head.value,
			})
		}

		next := std.head.next

		for next != nil {
			index++

			if next.value == val {
				resultSet = append(resultSet, map[string]interface{}{
					"idx": index,
					"val": next.value,
				})
			}

			next = next.next
		}

	}

	return resultSet
}

func (std *LinkedList) Update(oldVal interface{}, newVal interface{}) bool {
	result := false

	if std.head != nil {

		if std.head.value == oldVal {
			std.head.value = newVal

			result = std.head.value == newVal
		}

		if !result {
			next := std.head.next

			for next != nil {
				if next.value == oldVal {
					next.value = newVal

					result = next.value == newVal
					break
				}

				next = next.next
			}
		}

	}

	return result
}

func (std *LinkedList) Delete(val interface{}) bool {
	result := false

	if std.head != nil {

		if std.head.value == val {
			if std.head.next != nil {
				std.head = std.head.next
			} else {
				std.head = nil
			}

			result = true
		}

		if !result {
			prev := std.head
			next := std.head.next

			for next != nil {

				if next.value == val {
					if next.next != nil {
						prev.next = next.next
					} else {
						prev.next = nil
					}

					result = true
					break
				}

				prev = next
				next = next.next
			}
		}
	}

	return result
}

func (std *LinkedList) Size() int {
	var count []interface{}

	if std.head != nil {
		count = append(count, std.head)

		next := std.head.next

		for next != nil {
			count = append(count, std.head.value)
			next = next.next
		}
	}

	return len(count)
}
